Multiple definitions of the api/handlers in 1.0.0.4

### Describe the release item

I am currently looking into upgrading the Zenoh interface for PX4 to Zenoh-pico 1.0.0.4, upgrading it from Zenoh-pico 0.7. 

### Problem

There are multiple translation units in that project which includes `zenoh-pico.h` which itself includes `zenoh-pico/api/handlers.h`.
From `1.0` onwards, `zenoh-pico/api/handlers.h` implements the functions in the header file using macros.
This causes a link error as multiple definitions of the same functions exists. 

https://github.com/PX4/PX4-Autopilot/blob/b1b0032b8d7ea7ef7a0b7edfe5a9ed787acb3f9c/src/modules/zenoh/publishers/zenoh_publisher.hpp#L49C2-L49C24

https://github.com/PX4/PX4-Autopilot/blob/b1b0032b8d7ea7ef7a0b7edfe5a9ed787acb3f9c/src/modules/zenoh/zenoh.cpp#L45

**Error:**

```bash
/opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: src/modules/zenoh/libmodules__zenoh.a(zenoh_config.cpp.obj): in function `z_ring_handler_sample_check':
/home/alexis/work/PX4-Zenoh/src/modules/zenoh/zenoh-pico/include/zenoh-pico/api/handlers.h:157: multiple definition of `z_ring_handler_sample_check'; src/modules/zenoh/libmodules__zenoh.a(zenoh.cpp.obj):/home/alexis/work/PX4-Zenoh/src/modules/zenoh/zenoh-pico/include/zenoh-pico/api/handlers.h:157: first defined here
/opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: src/modules/zenoh/libmodules__zenoh.a(zenoh_config.cpp.obj): in function `z_ring_handler_sample_loan':
/home/alexis/work/PX4-Zenoh/src/modules/zenoh/zenoh-pico/include/zenoh-pico/api/handlers.h:157: multiple definition of `z_ring_handler_sample_loan'; src/modules/zenoh/libmodules__zenoh.a(zenoh.cpp.obj):/home/alexis/work/PX4-Zenoh/src/modules/zenoh/zenoh-pico/include/zenoh-pico/api/handlers.h:157: first defined here
/opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: src/modules/zenoh/libmodules__zenoh.a(zenoh_config.cpp.obj): in function `z_ring_handler_sample_loan_mut':
zenoh_config.cpp:(.text.z_ring_handler_sample_loan_mut+0x0): multiple definition of `z_ring_handler_sample_loan_mut'; src/modules/zenoh/libmodules__zenoh.a(zenoh.cpp.obj):zenoh.cpp:(.text.z_ring_handler_sample_loan_mut+0x0): first defined here
/opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: src/modules/zenoh/libmodules__zenoh.a(zenoh_config.cpp.obj): in function `z_ring_handler_sample_null':
/home/alexis/work/PX4-Zenoh/src/modules/zenoh/zenoh-pico/include/zenoh-pico/api/handlers.h:157: multiple definition of `z_ring_handler_sample_null'; src/modules/zenoh/libmodules__zenoh.a(zenoh.cpp.obj):/home/alexis/work/PX4-Zenoh/src/modules/zenoh/zenoh-pico/include/zenoh-pico/api/handlers.h:157: first defined here
/opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: src/modules/zenoh/libmodules__zenoh.a(zenoh_config.cpp.obj): in function `z_ring_handler_sample_move':
/home/alexis/work/PX4-Zenoh/src/modules/zenoh/zenoh-pico/include/zenoh-pico/api/handlers.h:157: multiple definition of `z_ring_handler_sample_move'; src/modules/zenoh/libmodules__zenoh.a(zenoh.cpp.obj):/home/alexis/work/PX4-Zenoh/src/modules/zenoh/zenoh-pico/include/zenoh-pico/api/handlers.h:157: first defined here
/opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: src/modules/zenoh/libmodules__zenoh.a(zenoh_config.cpp.obj): in function `z_ring_handler_sample_drop':
/home/alexis/work/PX4-Zenoh/src/modules/zenoh/zenoh-pico/include/zenoh-pico/api/handlers.h:157: multiple definition of `z_ring_handler_sample_drop'; src/modules/zenoh/libmodules__zenoh.a(zenoh.cpp.obj):/home/alexis/work/PX4-Zenoh/src/modules/zenoh/zenoh-pico/include/zenoh-pico/api/handlers.h:157: first defined here
/opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: src/modules/zenoh/libmodules__zenoh.a(zenoh_config.cpp.obj): in function `z_fifo_handler_sample_check':
/home/alexis/work/PX4-Zenoh/src/modules/zenoh/zenoh-pico/include/zenoh-pico/api/handlers.h:162: multiple definition of `z_fifo_handler_sample_check'; src/modules/zenoh/libmodules__zenoh.a(zenoh.cpp.obj):/home/alexis/work/PX4-Zenoh/src/modules/zenoh/zenoh-pico/include/zenoh-pico/api/handlers.h:162: first defined here
[This goes on, it triggers for all functions of handlers.h]
```

### Ideal solution

There shouldn't be any implementation in the header file to avoid breaking the **ODR (One Definition Rule)**.

### Temporary fix

It is possible to discard the error but it is not a solution.

```cmake
target_link_libraries(modules__zenoh  PRIVATE -Wl,--allow-multiple-definition)
```

@DenisBiryukov91 I created a minimal reproduction branch: https://github.com/AlexisTM/zenoh-pico/tree/issue/linking_odr

Compare: https://github.com/eclipse-zenoh/zenoh-pico/compare/release/1.0.0.4...AlexisTM:zenoh-pico:issue/linking_odr

Just building the example results in:

```bash
/usr/bin/ld: CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub_tu2.c.o: in function `z_ring_handler_sample_check':
/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub_tu2.c:157: multiple definition of `z_ring_handler_sample_check'; CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub.c.o:/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub.c:157: first defined here
/usr/bin/ld: CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub_tu2.c.o: in function `z_ring_handler_sample_loan':
/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub_tu2.c:157: multiple definition of `z_ring_handler_sample_loan'; CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub.c.o:/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub.c:157: first defined here
/usr/bin/ld: CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub_tu2.c.o: in function `z_ring_handler_sample_loan_mut':
/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub_tu2.c:157: multiple definition of `z_ring_handler_sample_loan_mut'; CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub.c.o:/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub.c:157: first defined here
/usr/bin/ld: CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub_tu2.c.o: in function `z_ring_handler_sample_null':
/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub_tu2.c:157: multiple definition of `z_ring_handler_sample_null'; CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub.c.o:/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub.c:157: first defined here
/usr/bin/ld: CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub_tu2.c.o: in function `z_ring_handler_sample_move':
/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub_tu2.c:157: multiple definition of `z_ring_handler_sample_move'; CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub.c.o:/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub.c:157: first defined here
/usr/bin/ld: CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub_tu2.c.o: in function `z_ring_handler_sample_drop':
/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub_tu2.c:157: multiple definition of `z_ring_handler_sample_drop'; CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub.c.o:/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub.c:157: first defined here
/usr/bin/ld: CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub_tu2.c.o: in function `z_fifo_handler_sample_check':
/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub_tu2.c:162: multiple definition of `z_fifo_handler_sample_check'; CMakeFiles/z_sub_tu2.dir/unix/c11/z_sub.c.o:/home/alexis/work/zenoh-pico/examples/unix/c11/z_sub.c:162: first defined here
[and so on]
```

In this case, it is not an embedded target, but a plain Unix, Ubuntu 24.04 using `gcc (Ubuntu 13.2.0-23ubuntu4) 13.2.0`.

Hope it helps

--- a/include/zenoh-pico/api/handlers.h
+++ b/include/zenoh-pico/api/handlers.h
@@ -23,10 +23,6 @@
 #include "zenoh-pico/collections/ring_mt.h"
 #include "zenoh-pico/utils/logging.h"
 
-#ifdef __cplusplus
-extern "C" {
-#endif
-
 // -- Channel
 #define _Z_CHANNEL_DEFINE_IMPL(handler_type, handler_name, handler_new_f_name, callback_type, callback_new_f,          \
                                collection_type, collection_new_f, collection_free_f, collection_push_f,                \
@@ -106,8 +102,8 @@ extern "C" {
         return h;                                                                                                      \
     }                                                                                                                  \
                                                                                                                        \
-    _Z_OWNED_FUNCTIONS_VALUE_NO_COPY_IMPL(handler_type, handler_name, _z_##handler_name##_check,                       \
-                                          _z_##handler_name##_null, _z_##handler_name##_clear)                         \
+    _Z_OWNED_FUNCTIONS_VALUE_NO_COPY_INLINE_IMPL(handler_type, handler_name, _z_##handler_name##_check,                \
+                                                 _z_##handler_name##_null, _z_##handler_name##_clear)                  \
                                                                                                                        \
     static inline int8_t handler_new_f_name(callback_type *callback, z_owned_##handler_name##_t *handler,              \
                                             size_t capacity) {                                                         \
@@ -191,7 +187,4 @@ _Z_CHANNEL_DEFINE_DUMMY(reply, ring)
 _Z_CHANNEL_DEFINE_DUMMY(reply, fifo)
 #endif  // Z_FEATURE_QUERY
 
-#ifdef __cplusplus
-}
-#endif
 #endif  // INCLUDE_ZENOH_PICO_API_HANDLERS_H

--- a/include/zenoh-pico/api/olv_macros.h
+++ b/include/zenoh-pico/api/olv_macros.h
@@ -113,16 +113,23 @@
     void z_##name##_drop(z_owned_##name##_t *obj) {                                                  \
         if (obj != NULL) f_drop((&obj->_val));                                                       \
     }
-
-#define _Z_OWNED_FUNCTIONS_VALUE_NO_COPY_IMPL(type, name, f_check, f_null, f_drop)                   \
-    _Bool z_##name##_check(const z_owned_##name##_t *obj) { return f_check((&obj->_val)); }          \
-    const z_loaned_##name##_t *z_##name##_loan(const z_owned_##name##_t *obj) { return &obj->_val; } \
-    z_loaned_##name##_t *z_##name##_loan_mut(z_owned_##name##_t *obj) { return &obj->_val; }         \
-    void z_##name##_null(z_owned_##name##_t *obj) { obj->_val = f_null(); }                          \
-    z_owned_##name##_t *z_##name##_move(z_owned_##name##_t *obj) { return obj; }                     \
-    void z_##name##_drop(z_owned_##name##_t *obj) {                                                  \
-        if (obj != NULL) f_drop((&obj->_val));                                                       \
+#define _Z_OWNED_FUNCTIONS_VALUE_NO_COPY_IMPL_INNER(type, name, f_check, f_null, f_drop, attribute)            \
+    attribute _Bool z_##name##_check(const z_owned_##name##_t *obj) { return f_check((&obj->_val)); }          \
+    attribute const z_loaned_##name##_t *z_##name##_loan(const z_owned_##name##_t *obj) { return &obj->_val; } \
+    attribute z_loaned_##name##_t *z_##name##_loan_mut(z_owned_##name##_t *obj) { return &obj->_val; }         \
+    attribute void z_##name##_null(z_owned_##name##_t *obj) { obj->_val = f_null(); }                          \
+    attribute z_owned_##name##_t *z_##name##_move(z_owned_##name##_t *obj) { return obj; }                     \
+    attribute void z_##name##_drop(z_owned_##name##_t *obj) {                                                  \
+        if (obj != NULL) f_drop((&obj->_val));                                                                 \
     }
+
+#define _ZP_NOTHING
+
+#define _Z_OWNED_FUNCTIONS_VALUE_NO_COPY_IMPL(type, name, f_check, f_null, f_drop) \
+    _Z_OWNED_FUNCTIONS_VALUE_NO_COPY_IMPL_INNER(type, name, f_check, f_null, f_drop, _ZP_NOTHING)
+
+#define _Z_OWNED_FUNCTIONS_VALUE_NO_COPY_INLINE_IMPL(type, name, f_check, f_null, f_drop) \
+    _Z_OWNED_FUNCTIONS_VALUE_NO_COPY_IMPL_INNER(type, name, f_check, f_null, f_drop, static inline)
 
 #define _Z_OWNED_FUNCTIONS_RC_IMPL(name)                                                            \
     _Bool z_##name##_check(const z_owned_##name##_t *val) { return val->_rc.in != NULL; }           \

